Sprint 19

medvedevda opened the pull request:
Выполнен спринт 19:
- сверстаны все экраны
- добавлена поддержка темной темы
- добавлена локализация всех экранов русского и английского языков
- сверстан webView для просмотра пользовательского соглашения

Единственный момент, при открытии webView у меня постоянно обновляется progressView. Делал все по учебнику с созданием Coordinator. Не могу понять в чем проблема, очень хотел бы получить подсказки и комментарии по поводу этого, где и что я не так делаю?

Спасибо

Commit:
Добавлен Splash Screen

Commit:
Сверстан TabView

- добавлены цвета для верстки UI с поддержкой темной темы
- подготовлены файлы для верстки главного экрана и экрана настроек

Commit:
Сверстан экран настроек, добавлена темная тема

Commit:
Сверстан главный экран с расписанием

исправлен launchScreen (сверстан через storyboard)

Commit:
Добавлен переход на экран выбора города отправления

исправлена ошибка, из-за которой отображался другой цвет в textField при смене темной темы

Commit:
Сверстан экран выбора станции

Commit:
Добавлена навигация между экранами выбора города и станции

- добавлена возможность поменять местами станции отправления и прибытия
- добавлен переход на экран с результатами поиска расписания

Commit:
Сверстан экран с детальной информацией о перевозчике

добавлен переход на экран с информацией о перевозчике
исправлены ошибки с отображением элементов при включении темной темы

Commit:
Сверстан экран с фильтрами

добавлена навигация между экранами результатов поиска расписания и экраном с фильтрами

Commit:
Сверстан экран WebView с пользовательским соглашением

- добавлена локализация (поддерживаются английский и русский языки)

Commit:
Сверстаны экраны для ситуации "Нет интернета" и "Ошибка сервера", локализованы и адаптированы под темную тему

Commit:
fix

Commit:
Вынесены константы ui-элементов в отдельный файл

Commit:
добавлена анимация для кнопки смены направления

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    class Coordinator: NSObject, WKNavigationDelegate {
+        var webView: WKWebView?
```
Обрати внимание, что в `Coordinator` мы указываем view-обёртку (т.е. `WebView`).

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+
+struct WebView: UIViewRepresentable {
+    let urlString: String
+    @Binding var progress: Double
```
Проблема с обновлением прогресса загрузки связана с тем, что сейчас происходит одновременное изменение значения свойства `WebView` (`progress`) и рендеринг самой вью `WebView`. 

Чтобы решить эту проблему, можно в `WebView` создать простую вью-модель, в которой будет храниться `progress`:

```
extension WebView {
    class ProgressViewModel: ObservableObject {
        @Published var progress: Double = 0.0
        
        init(progress: Double) {
            self.progress = progress
        }
    }
}
```

И в WebView вместо `@Binding var progress: Double` использовать эту вью-модель:
`@ObservedObject var viewModel: ProgressViewModel`

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+    @Binding var progress: Double
+    
+    func makeUIView(context: Context) -> WKWebView {
+        let webView = WKWebView()
```
`WKWebView` нужно вынести в приватную константу, т.к. к ней понадобится доступ из `Coordinator`:
`private let webView = WKWebView()`

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+    func makeUIView(context: Context) -> WKWebView {
+        let webView = WKWebView()
+        webView.navigationDelegate = context.coordinator
+        webView.addObserver(context.coordinator,
```
Observer мы вынесем в `Coordinator`, поэтому здесь нужно удалить его добавление. 
В принципе, в `func makeUIView(context: Context)` достаточно оставить только загрузку в `webView`:
```
if let url = URL(string: urlString) {
            webView.load(URLRequest(url: url))
        }
        return webView
```

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    func updateUIView(_ uiView: WKWebView, context: Context) {
+        if let url = URL(string: urlString) {
```
В `func updateUIView(_ uiView: WKWebView, context: Context)` мы ничего не будем делать, т.к. обновление и данных будет выполнять через observer, поэтому здесь тело функции оставляем пустым.

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+    
+    class Coordinator: NSObject, WKNavigationDelegate {
+        var webView: WKWebView?
+        @Binding var progress: Double
```
В Coordinator будем хранить 3 свойства:
```
private var parent: WebView
private var viewModel: ProgressViewModel
private var observer: NSKeyValueObservation?
```

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+        var webView: WKWebView?
+        @Binding var progress: Double
+        
+        init(webView: WKWebView? = nil, progress: Binding<Double>) {
```
С учетом изменений выше, здесь нужно поменять инициализатор (как раз здесь будем устанавливать observer - нужно не забыть в `deinit` удалить его):

```
init(parent: WebView, viewModel: ProgressViewModel) {
            self.parent = parent
            self.viewModel = viewModel
            super.init()
            
            self.observer = self.parent.webView.observe(\.estimatedProgress) { [weak self] webView, _ in
                guard let self else { return }
                self.parent.viewModel.progress = parent.webView.estimatedProgress
            }
        }
```

jshlykova commented on TravelSchedule/View/WebView/WebView.swift:
```diff
@@ -0,0 +1,53 @@
+            webView?.removeObserver(self, forKeyPath: #keyPath(WKWebView.estimatedProgress))
+        }
+        
+        override func observeValue(forKeyPath keyPath: String?,
```
Можно удалить, т.к. теперь храним отдельно `private var observer: NSKeyValueObservation?`

jshlykova commented on TravelSchedule/View/WebView/AgreementView.swift:
```diff
@@ -0,0 +1,20 @@
+import SwiftUI
+
+struct AgreementView: View {
+    @State private var progress: Double = 0.0
```
И здесь тоже вместо progress теперь используем созданную вью-модель:
`@StateObject var viewModel = WebView.ProgressViewModel(progress: 0.0)`

(соответственно, ниже нужно поменять код с учетом вью-модели)

jshlykova reviewed (commented):
Дмитрий, привет!
Проверила твою работу.
Оставила комментарии по поводу исправления проблемы с отображением прогресса загрузки. 
Твой подход верный и он соответствует описанию в учебнике. Просто в учебнике пример более простой, а с прогрессом ситуация посложнее, поэтому вопрос вполне закономерный и очень здорово, что ты его задал. 
Надеюсь, комментарии помогут разобраться и решить проблему.

Commit:
Исправлена ошибка с обновлением progressView в webView.

- добавлен navigationTitle и кастомная кнопка назад в ToolBar на экране пользовательского соглашения
- исправлены горизонтальные отступы у изображения перевозчика на экране CarrierInfoView

Commit:
Исправлена ошибка, связанная с обработкой нажатия на главном экране (в поле выбора "откуда" и "куда")

D1mon227 commented:
Исправил ошибку, проблема не возникает. Спасибо
Но теперь появляется сообщение по поводу UI (Publishing changes from within view updates is not allowed, this will cause undefined behavior.) Почитал об этом: как я понял, это возникает из-за того, что изменения в модели данных происходят внутри замыканий, вызванных во время обновления представления.
Также в deinit если удалять observer через метод removeObserver, то когда выхожу из webView на экран настроек - то приложение падает с ошибкой "не удалось удалить observer, так как он не зарегистрирован как observer". Мы его добавили через .observe(.estimatedProgress), а не через метод .addObserver.

Может быть где-то я все-таки ошибся и неправильно понял? Еще наткнулся в интернете, что можно поменять нашу viewModel в webView на @StateObject вместо @ObservedObject, но проблема остается.

jshlykova commented on TravelSchedule/View/AgreementView/WebView.swift:
```diff
@@ -0,0 +1,52 @@
+            
+            self.observer = self.parent.webView.observe(\.estimatedProgress) { [weak self] webView, _ in
+                guard let self else { return }
+                self.parent.viewModel.progress = parent.webView.estimatedProgress
```
Проблема `Publishing changes from within view updates is not allowed, this will cause undefined behavior` возникает из-за того, что сейчас `WebView` обновляет `@Published`-свойство `progress`.

Должен помочь перенос операции изменения значения в главную очередь:
```
DispatchQueue.main.async { 
                    self.parent.viewModel.progress = parent.webView.estimatedProgress
                }
```

medvedevda replied:
Да, на stackOverFlow тоже нашел ответ, что да, нужно перенести в главную очередь. Попробую.

jshlykova reviewed (commented):
Дмитрий, привет!

Проверила твою работу. 
По поводу ошибки написала в [комментарии](https://github.com/D1mon227/TravelSchedule/pull/2/files#r1590011344).

Commit:
Исправлена ошибка, связанная с обработкой нажатия на экранах выбора города и станции

исправлена ошибка "Publishing changes from within view updates is not allowed, this will cause undefined behavior." в экране пользовательского соглашения
